x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/privilege/TransportClearPrivilegesCacheAction.java

```diff
@@ -0,0 +1,82 @@
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.security.action.privilege.ClearPrivilegesCacheAction;
+import org.elasticsearch.xpack.core.security.action.privilege.ClearPrivilegesCacheRequest;
+import org.elasticsearch.xpack.core.security.action.privilege.ClearPrivilegesCacheResponse;
+import org.elasticsearch.xpack.security.authz.store.CompositeRolesStore;
+import org.elasticsearch.xpack.security.authz.store.NativePrivilegeStore;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+public class TransportClearPrivilegesCacheAction extends TransportNodesAction<ClearPrivilegesCacheRequest, ClearPrivilegesCacheResponse,
+    ClearPrivilegesCacheRequest.Node, ClearPrivilegesCacheResponse.Node> {
+
+    private final NativePrivilegeStore privilegesStore;
+    private final CompositeRolesStore rolesStore;
+
+    @Inject
+    public TransportClearPrivilegesCacheAction(
+        ThreadPool threadPool,
+        ClusterService clusterService,
+        TransportService transportService,
+        ActionFilters actionFilters,
+        NativePrivilegeStore privilegesStore,
+        CompositeRolesStore rolesStore) {
+        super(
+            ClearPrivilegesCacheAction.NAME,
+            threadPool,
+            clusterService,
+            transportService,
+            actionFilters,
+            ClearPrivilegesCacheRequest::new,
+            ClearPrivilegesCacheRequest.Node::new,
+            ThreadPool.Names.MANAGEMENT,
+            ClearPrivilegesCacheResponse.Node.class);
+        this.privilegesStore = privilegesStore;
+        this.rolesStore = rolesStore;
+    }
+
+    @Override
+    protected ClearPrivilegesCacheResponse newResponse(
+        ClearPrivilegesCacheRequest request, List<ClearPrivilegesCacheResponse.Node> nodes, List<FailedNodeException> failures) {
+        return new ClearPrivilegesCacheResponse(clusterService.getClusterName(), nodes, failures);
+    }
+
+    @Override
+    protected ClearPrivilegesCacheRequest.Node newNodeRequest(ClearPrivilegesCacheRequest request) {
+        return new ClearPrivilegesCacheRequest.Node(request);
+    }
+
+    @Override
+    protected ClearPrivilegesCacheResponse.Node newNodeResponse(StreamInput in) throws IOException {
+        return new ClearPrivilegesCacheResponse.Node(in);
+    }
+
+    @Override
+    protected ClearPrivilegesCacheResponse.Node nodeOperation(ClearPrivilegesCacheRequest.Node request, Task task) {
+        if (request.getApplicationNames() == null || request.getApplicationNames().length == 0) {
+            privilegesStore.invalidateAll();
+        } else {
+            privilegesStore.invalidate(Arrays.asList(request.getApplicationNames()));
+        }
+        rolesStore.invalidateAll();
```

Comment:
I'm not sure about this. It seems like this API ends up doing something other than what it was supposed to, just because we assume that the caller wants it.
I understand why - if the privileges have changed then the roles cache is probably wrong, but it seems like it's chain side-effects together.

Reply:
From just the API point of view, you are right that these two should not be tied together. There are valid use cases when user only wants to actively clear privileges cache. I did this because the two are always tied together in `NativePrivilegesStore` since it was clearing role cache before my change.

I tried to avoid nested callbacks (clear role cache then clear privileges cache). It seems OK and more efficient by just looking at `NativePrivilegesStore`. But it does feel wrong from pure API side.

I could either just go with nested callback or create a transport layer only action to clear both caches. So it is not exposed at REST layer and still has the efficiency on transport layer. But this does lead some code redundancy. Another option is to have a query parameter for the clear privilege cache API. When set to true, it clears both caches. What do you think?